pkg/cache/userpermission.go

```diff
@@ -0,0 +1,1107 @@
+	return discoverableRoles, nil
+}
+
+// getPermissionRecords returns the aggregated permissions from user and groups
+func (c *UserPermissionCache) getPermissionRecords(userInfo user.Info) map[string][]clusterviewv1alpha1.ClusterBinding {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	// Map of roleName -> []ClusterBinding
+	roleBindings := make(map[string][]clusterviewv1alpha1.ClusterBinding)
+	userName := userInfo.GetName()
+	groups := userInfo.GetGroups()
+
+	// Get permissions from user
+	obj, exists, _ := c.userPermissionStore.GetByKey(userName)
+	if exists {
+		record := obj.(*UserPermissionRecord)
+		for _, perm := range record.Permissions {
+			for _, binding := range perm.Bindings {
+				roleBindings[perm.ClusterRoleName] = mergeOrAppendBinding(roleBindings[perm.ClusterRoleName], binding)
+			}
+		}
+	}
+
+	// Get permissions from groups
+	for _, group := range groups {
+		obj, exists, _ := c.groupPermissionStore.GetByKey(group)
+		if exists {
+			record := obj.(*UserPermissionRecord)
+			for _, perm := range record.Permissions {
+				for _, binding := range perm.Bindings {
+					roleBindings[perm.ClusterRoleName] = mergeOrAppendBinding(roleBindings[perm.ClusterRoleName], binding)
+				}
+			}
+		}
+	}
+
+	return roleBindings
+}
+
+// List returns the list of UserPermissions for a user
+func (c *UserPermissionCache) List(
+	userInfo user.Info, selector labels.Selector,
+) (*clusterviewv1alpha1.UserPermissionList, error) {
+	roleBindings := c.getPermissionRecords(userInfo)
+
+	userPermissionList := &clusterviewv1alpha1.UserPermissionList{
+		Items: make([]clusterviewv1alpha1.UserPermission, 0, len(roleBindings)),
+	}
+
+	c.lock.RLock()
+	discoverableRoles := c.discoverableRoles
+	c.lock.RUnlock()
+
+	for roleName, bindings := range roleBindings {
+		if len(bindings) == 0 {
+			continue
+		}
+
+		userPerm := clusterviewv1alpha1.UserPermission{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: roleName,
+			},
+			Status: clusterviewv1alpha1.UserPermissionStatus{
+				Bindings: bindings,
+			},
+		}
+
+		// Add ClusterRole definition
+		switch roleName {
+		case clusterviewv1alpha1.ManagedClusterAdminRole:
+			userPerm.Status.ClusterRoleDefinition = c.getSyntheticAdminRoleDefinition()
+		case clusterviewv1alpha1.ManagedClusterViewRole:
+			userPerm.Status.ClusterRoleDefinition = c.getSyntheticViewRoleDefinition()
+		default:
+			// Look for the role in discoverable roles
+			for _, role := range discoverableRoles {
+				if role.Name == roleName {
+					userPerm.Status.ClusterRoleDefinition = clusterviewv1alpha1.ClusterRoleDefinition{
+						Rules: role.Rules,
+					}
+					break
+				}
+			}
+		}
+
+		userPermissionList.Items = append(userPermissionList.Items, userPerm)
+	}
+
+	return userPermissionList, nil
+}
+
+// Get returns a specific UserPermission by ClusterRole name
+func (c *UserPermissionCache) Get(userInfo user.Info, name string) (*clusterviewv1alpha1.UserPermission, error) {
```

Comment:
I think we should add rlock around this get. We do not need to lock in private method separately.